### ISSUE TITLE
Allow specifying custom PowerShell config and cache directory instead of the hardcoded paths in `Platform`

### ISSUE THREAD

**Issue by MatejKafka** [PowerShell/PowerShell#13815]
## Summary of the new feature/enhancement
Currently, it is possible to download PowerShell Core as a .zip archive and unzip it to a custom directory and everything seems to work correctly even from non-standard path, so users could run `pwsh` as a portable package by putting it e.g. on a USB flash drive and just plugging it in whenever needed. 

However, the config directory (where profile and installed modules are stored; at `Documents/Powershell`) and cache directory (at `LocalAppData\Microsoft\PowerShell`) paths are currently hardcoded. It would be great if we could specify custom paths in `powershell.config.json` or another similar config file, ideally with support for paths relative to the `pwsh.exe` binary.

## Proposed technical implementation details (optional)
It seems that both directories are configured here: https://github.com/PowerShell/PowerShell/blob/1a3721a574e00263f64101d5dbbe1a975e6c8087/src/System.Management.Automation/CoreCLR/CorePsPlatform.cs#L149

As far as I can tell, it should be enough to check if there are any paths configured in the config file and set them here (and ideally abort with an error if the paths do not exist). I don't have much experience with C#, but I can try writing a pull request if we agree on a solution.

**Comment by iSazonov** [comment]
Our current intention is to exclude config file at all and use defaults because parsing a config file slows down startup scenario noticeably. 

As for `Documents/Powershell`, you can set PSModulePath environment variable as needed at startup.
As for cache directory, we follow OS standards.

We had issues with Docker scenario and fixed them but portable scenario works well as-is.
To approve the enhancement we need to see _an important scenario that is blocked_.

**Comment by fMichaleczek** [comment]
> a config file slows down startup scenario noticeable
> As for `Documents/Powershell`, you can set PSModulePath environment variable
> To approve the enhancement we need to see _an important scenario that is blocked_.
Portable behavior on Windows is a kind of packaging that need a clean software and PowerShell is not. ☹️ It's a shame for us. Portable community always exists and if they need an environment variable like PSModuleCachePath, we should create one.

**Comment by MatejKafka** [comment]
> We had issues with Docker scenario and fixed them but portable scenario works well as-is.
> To approve the enhancement we need to see _an important scenario that is blocked_.

For me personally, I'm working on a package manager and packaging system for Windows that partially mirrors the macOS package structure (each app has its own directory, and all binaries, config, data, logs,... are stored in this directory). This allows me to reinstall my system and keep all my apps in the same state without needing to reinstall them.

Another useful scenario: I'm a sysadmin, and I carry an USB flash drive which contains utilities that I use for fixing and managing machines, including `pwsh`. This allows me to just connect the drive wherever needed and instantly, I have all my tools ready without installation and configuration.

Given how minor the requested change is, I personally feel it enables scenarios useful enough to be worth it.

**Comment by MatejKafka** [comment]
> Portable behavior on Windows is a kind of packaging that need a clean software and PowerShell is not. frowning_face It's a shame for us. Portable community always exists and if they need an environment variable like PSModuleCachePath, we should create one.

I don't feel an extra environment variable is a good solution - for the portable scenario, it needs an extra launcher just to pass an environment variable, as we cannot modify global env on the target machine.

**Comment by fMichaleczek** [comment]
@MatejKafka Yes, I know but I already  use environment variable to exclude TELEMETRY, so ...

There is already a PSModuleAnalysisCachePath environment variable, I personally disable the analysisCache on portable scenario and set, PSDisableModuleAnalysisCacheCleanup to 1. (the first launch frozen in the past to build the module cache, I idk it's still the same). The documentation doesn't explain if it's also cover the config file. 
[About PowerShell Config](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-7.1)

I think both should be in the powershell.config.json

**Comment by dweddig01** [comment]
> As for `Documents/Powershell`, you can set PSModulePath environment variable as needed at startup.

Unfortunately, we are experiencing behavior where the PSModulePath we set, whether in Windows System Environment Variables or in the startup profile, gets the 'documents/powershell' added to the front of the variable. This happens everytime we use Install-Module or Update-Module. I am not sure why that happens. There should be logic that allows you to set that PSModulePath and not have it adjusted if it is set either at the system, or within the profile startup.

**Comment by iSazonov** [comment]
@dweddig01 Could you please open new issue and describe your business scenario?

**Comment by MatejKafka** [comment]
not completed, please reopen

**Comment by MatejKafka** [comment]
@kilasuit not completed, please reopen

**Comment by MatejKafka** [comment]
Still relevant.

**Comment by MatejKafka** [comment]
Still relevant.

**Comment by smartground** [comment]
And still relevant
